### PR TITLE
Time entry model

### PR DESCRIPTION
**Pull request by simonrozsival**

### CODE REVIEW

**Comment by simonrozsival** [comment]
I just realized that I ran the tests only for `Toggl.Ultrawave` and not for other projects, so I didn't notice, that `IDatabaseTimeEntry` extends `ITimeEntry` and that some projects in the solution won't build. Should I implement the database stuff as well in this PR or should I rather remove the impmentation of `ITimeEntry` for now?

**Comment by amulware** [comment]
Since you are only adding the model and nothing else, I'm fine if you add the database model too, no problem there.

**Comment by simonrozsival** [comment]
I stared reusing #65 but it seems that we can't use the generic Lists for tags and tag ids - build fails with `Fody/RealmWeaver [...] generic list is not yet supported`.

**Comment by simonrozsival** [comment]
Well I was a too impetuous with the previous commits - after reading the documentation of Realm I found out that even though the `System.Collections.Generic.List` class is not supported, the interface `System.Collections.Generic.IList` supported is and it works well with the serialization/deserialization as well.

File: Toggl.Foundation/Models/FoundationModels.cs [hunk 1]
```diff
@@ -46,6 +47,34 @@ internal partial class TimeEntry : IDatabaseTimeEntry
...
+
+        public IList<string> Tags { get; }
+
+        public IList<int> TagIds { get; }
+
+        public DateTimeOffset At { get; }
+
+        public DateTimeOffset? ServerDeletedAt { get; }
+
+        public int Uid { get; }
```
Comment (willsb) [inline Toggl.Foundation/Models/FoundationModels.cs]: Could use a better name

Comment (simonrozsival) [inline Toggl.Foundation/Models/FoundationModels.cs]: You are right, but I have no clue what "uid" stands for..

Comment (refiito) [inline Toggl.Foundation/Models/FoundationModels.cs]: Uid - UserID, use the JSON field of `user_id` (`uid` is the same, but basically for webapp backwards compatibility)

Comment (willsb) [inline Toggl.Foundation/Models/FoundationModels.cs]: This means that you should also change the mappings in the Ultrawave layer to reflect this (no more `wid` or `pid` there)

Comment (simonrozsival) [inline Toggl.Foundation/Models/FoundationModels.cs]: OK, I reused some older already reviewed code from a different PR where there was both Uid and UserId - I will remove uid then.

**Comment by simonrozsival** [comment]
I have run and fixed the t4 templates by now, so that should be OK.

File: Toggl.Foundation/Models/FoundationConstructors.cs [hunk 2]
```diff
@@ -1,4 +1,8 @@
-﻿﻿using Toggl.Multivac.Models;
+﻿
+
+
+
+using Toggl.Multivac.Models;
```
Comment (amulware) [inline Toggl.Foundation/Models/FoundationConstructors.cs]: Empty lines?

File: Toggl.Foundation/Models/FoundationModels.cs [hunk 3]
```diff
@@ -1,4 +1,9 @@
-﻿﻿using System;
+﻿
+
+
+
+using System;
```
Comment (amulware) [inline Toggl.Foundation/Models/FoundationModels.cs]: More empty lines?

File: Toggl.PrimeRadiant.Realm/Models/RealmConstructors.cs [hunk 4]
```diff
@@ -1,4 +1,8 @@
-﻿﻿using Realms;
+﻿
+
+
+
+using Realms;
```
Comment (amulware) [inline Toggl.PrimeRadiant.Realm/Models/RealmConstructors.cs]: Empty lines?

File: Toggl.Foundation/Models/FoundationConstructors.cs (hunk 2 above)
Comment (simonrozsival) [inline Toggl.Foundation/Models/FoundationConstructors.cs]: That file is generated by t4, maybe @willsb will know how to eliminate those?

File: Toggl.Foundation/Models/FoundationModels.cs (hunk 3 above)
Comment (simonrozsival) [inline Toggl.Foundation/Models/FoundationModels.cs]: Same here

File: Toggl.PrimeRadiant.Realm/Models/RealmConstructors.cs (hunk 4 above)
Comment (simonrozsival) [inline Toggl.PrimeRadiant.Realm/Models/RealmConstructors.cs]: Same here

File: Toggl.Foundation/Models/FoundationModels.tt [hunk 5]
```diff
@@ -1,8 +1,9 @@
-﻿﻿<#@ template language="C#" hostspecific="True" #>
-<#@ assembly name="System.Core" #>
-<#@ import namespace="System.IO" #>
-<#@ import namespace="System.Linq" #>
-using System;
+﻿﻿<#@ template language="C#" hostspecific="True"
+#><#@ assembly name="System.Core"
```
Comment (simonrozsival) [inline Toggl.Foundation/Models/FoundationModels.tt]: I don't really like this solution, but it works well. This way it does not generate the empty lines at the start of the generated file.

File: Toggl.Ultrawave/Models/TimeEntry.cs [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+using Toggl.Multivac.Models;
+using System.Collections.Generic;
+
+namespace Toggl.Ultrawave
+{
+    public sealed class TimeEntry : ITimeEntry
+    {
+        public int Id { get; set; }
+
+        [JsonProperty("wid")]
```
Comment (willsb) [inline Toggl.Ultrawave/Models/TimeEntry.cs]: Quoting Margus in mobile dev:
> tl;dr; don’t use `uid`, `wid`, `pid`, etc… use `user_id`, `workspace_id`, `project_id` and so forth

This should change, I believe.

Comment (simonrozsival) [inline Toggl.Ultrawave/Models/TimeEntry.cs]: OK

Comment (simonrozsival) [inline Toggl.Ultrawave/Models/TimeEntry.cs]: I removed the annotations and changed the test.

**Comment by willsb** [comment]
:shipit:
